display_lists: minor compile-time op size tweaks

gaaclarke opened the pull request:
This has no effect on runtime behavior.  It just makes the following compile-time changes:
1) Makes the size of the op type 1 byte to match the size in DlOp (This really should have been a compilation error).
1) ~~Converts all the size comments to static_asserts, updates them or deletes them when they were incorrect.~~

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

gaaclarke commented:
Actually, this PR might fail CI since the sizes may vary depending on the platform and std implementation.  Let's see.

flar commented on display_list/display_list.h:
```diff
@@ -138,14 +138,16 @@ namespace flutter {
 
 #define DL_OP_TO_ENUM_VALUE(name) k##name,
-enum class DisplayListOpType {
+enum class DisplayListOpType : uint8_t {
```
The only place this is used, I believe, (at least used in a place where we care about the size) already has a size specifier on it to keep it to 8 bits.

gaaclarke replied:
Yep, it's marked as 8 bits in the DlOp.  C++ isn't complaining about throwing away the bits.  This is safer since we won't be throwing away bits anymore.

gaaclarke replied:
Without this change sizeof(DisplayListOpType) is 4.  I guess because it's a class enum though it may be able to determine that throwing away those bits isn't an issue.

flar replied:
Right, we don't have an assert that we are using less than 256 values. I guess I don't have an issue, but more that I'm wondering why we are worried about this all of a sudden.

gaaclarke replied:
You mean why do I bother to read the engine code critically?  =)

gaaclarke replied:
Hmm, looks like this broke some tests.  I'm guessing somewhere in the test code we were using sizeof(DisplayListOpType) instead of sizeof(DlOp)?  I can't see why, seems like a bug in testing code, but it isn't apparent where.  If it isn't obvious to you we can drop this.

flar replied:
> You mean why do I bother to read the engine code critically? =)

No, that's a good thing, I'm referring more to priority.

flar replied:
> Hmm, looks like this broke some tests. I'm guessing somewhere in the test code we were using sizeof(DisplayListOpType) instead of sizeof(DlOp)? I can't see why, seems like a bug in testing code, but it isn't apparent where. If it isn't obvious to you we can drop this.

It looks like it's mainly Windows which I've always had trouble with. Given that this change causes a huge increase in the storage sizes (as seen in the output of the results of the SingleOp tests), I wonder if the problem is how it applies the bitfield specifiers on this line: https://github.com/flutter/engine/blob/f9c497f178d3a8a8b77731c8a15ea46b558b0081/display_list/dl_op_records.h#L102

Perhaps it will bitfield-pack an enum into 8 bits, but when the type inherits from `uint8_t` then it decides to break the bitfields apart so before it had 8 bits of type and 24 bits of size == 32 bits, but now it has "uh, oh, that's a uint8_t, I can't pack that so I'll give it its own byte" followed by "and now they want a size field packed into 24 bits? Well that's an entirely new thing so I'll align this part to the next word boundary". I could run some experiments with godbolt and our Windows compilers to see if they suddenly refuse to pack the fields. I've seen similar cases in the past where it makes pessimal assumptions about packing that seem totally unnecessary but it ends up wasting a lot of bytes in the structure packing.

The ordering of the fields in the ops were designed to eliminate as much wasted packing space as they could so suddenly having the combined type/size bitfields expand from 4 to 8 bytes means that the ordering of the fields in the op records after that might break. Typically I'll try to have the first thing in an op record by a 32-bit quantity so that if it is followed by a 64-bit (or longer) aligned field then we'll use those 32-bits rather than skip them. If the end of the type/size bits is now 64-bit aligned then all of the alignments after that will be out of phase...

flar replied:
One could also ask if all of this size testing is really necessary, or if it is a premature micro-optimization, but this tiny change points out that what looks like "good C++ code" ends up causing Windows to waste a bunch of bits in the structs...

flar commented:
> Actually, this PR might fail CI since the sizes may vary depending on the platform and std implementation. Let's see.

Windows does not play well with the size estimates. Also, many of the fields vary in size for 32-bit and 64-bit pointers.

flar commented:
> * Makes the size of the op type 1 byte to match the size in DlOp (This really should have been a compilation error).

I'm not sure why this is an issue or why it should have been a compilation error...?

gaaclarke commented:
> Windows does not play well with the size estimates. Also, many of the fields vary in size for 32-bit and 64-bit pointers.

Yea, i reverted it.  I'm not sure what we want to do.  Some of the comments were way off.  We could just static assert for one platform to have a baseline and a ballpark estimate.  I just reverted for now.

flar commented:
Another direction to take would be protective asserts or FML_CHECKS that assert that we can fit all of the enum values into the `type` bitfield. Something along the lines of `FML_CHECK(op->type == T::Type)` in the `Push` method to verify that the value we read back from the field is the value we tried to put there. Or, static assert that the max enum value fits in the number of bits we assign to the field?

chinmaygarde commented:
Are we making progress on this? The tests are failing.

chinmaygarde commented:
Added the WIP tag. But please close if no progress is being made.

gaaclarke commented:
> Added the WIP tag. But please close if no progress is being made.

I don't think it's worth spending any more time on considering the weird behavior from the windows compiler.
